### DISCUSSION TITLE
LED staying on with asyncio

### DISCUSSION THREAD

**Discussion by SandyWalsh** [micropython/micropython#16016] [RP2040 / Pico] [answered]
```
MPY: soft reboot
MicroPython v1.23.0 on 2024-06-02; Raspberry Pi Pico with RP2040

>>> os.uname()
(sysname='rp2', nodename='rp2', release='1.23.0', version='v1.23.0 on 2024-06-02 (GNU 13.2.0 MinSizeRel)', machine='Raspberry Pi Pico with RP2040')
```

My program
```
from machine import Pin
import asyncio

async def blink(led, period_ms, evt):
    try:
        print("START")
        while not evt.is_set():
            print("-ON")
            led.on()
            await asyncio.sleep_ms(5)
            led.off()
            print("-OFF")
            await asyncio.sleep_ms(period_ms)
        print("EXIT")
    except:
        print("ERROR")
        for x in range(5):
            led.on()
            await asyncio.sleep(100)
            led.off()
            await asyncio.sleep(100)

async def main(led):
    event=asyncio.Event()
    print("$ MAIN")
    t=asyncio.create_task(blink(led, 700, event))
    print("$ SLEEPING")
    await asyncio.sleep(3)
    print("$ SETTING FLAG")
    event.set()
    print("$ WAITING")
    await asyncio.wait_for(t, 1000)
    print("$ DONE")

led=Pin(25, Pin.OUT, value=0)
asyncio.run(main(led))
```

The output looks correct (ends with a "-OFF" message)

```
>>> asyncio.run(main(led))
$ MAIN
$ SLEEPING
START
-ON
-OFF
-ON
-OFF
-ON
-OFF
-ON
-OFF
-ON
-OFF
$ SETTING FLAG
$ WAITING
EXIT
$ DONE
```

But the LED is staying on about 50% of the time. If I explicitly do a `finally: led.off()` it's fine, but that shouldn't be necessary. 

Thoughts?

**Comment by SandyWalsh** [comment]
On further investigation it appears that some of the `led.on()` calls aren't making it through either. I would expect to see 5 flashes, but many times I just see 4. This could explain why there are times when the led is left on.

**Comment by SandyWalsh** [comment]
Created issue for this 
https://github.com/micropython/micropython/issues/16019

**Reply by peterhinch** [reply]
Please don't. There is no evidence of a MicroPython bug here.

**Comment by glenn20** [comment]
I have no definitive answer, but if I was trying to debug this, I would try changing the 5ms delay before turning off the LED to 100 ms ( or more) in case there is some timing or race condition. Asyncio granularity isn't well suited to precise timing at the ms level.

**Reply by GitHubsSilverBullet** [reply]
Have modified it slighly:
```py
led=Pin(25, Pin.OUT, value=0)
while True:
    asyncio.run(main(led))
    if led.value():
        raise ValueError
```

It's been running for quite a while now, showing no signs of a problem.
Are you using – dare I say it – Thonny?

**Comment by jomasnash** [comment]
If I run this code in Thonny (rp2040 pico) with led 5ms on, I always get 5 blinks and the led is always off when the program is finished

**Comment by SandyWalsh** [comment]
Thanks for trying this everyone. Some additional information:

1. I originally saw the problem running like this:
```
$ picocom --baud 115200 /dev/ttyUSB0
```
and copy-pasting the script

2. I've also tried this with 
```
$ mpremote connect /dev/ttyUSB0 run ./raw_pico.py 
```
but didn't see the problem. 

3. When I added the `led.value()` checks I don't see the problem. 

4. When I added a loop to continuously do the test I didn't see the problem. 

The only thing I can imagine is there was some odd cruft between runs using the `picocom` approach? I wasn't doing a soft-reset between runs. 

Perhaps there were old coroutines kicking around? 

I'll continue to try to get faithful reproduction steps.

**Reply by corranwebster** [reply, marked as answer]
> The only thing I can imagine is there was some odd cruft between runs using the picocom approach? I wasn't doing a soft-reset between runs.

This seems the most likely cause: each time you run without reset you are creating a new event/task pair and there may be some code path where the old tasks persist (eg. if they did hit the timeout, then they may remain active as tasks, not 100% sure of the implementation internals here) and so are run by the event loop the next time you start it up alongside your new code.

This is a difference between Micropython and Python - Python Asyncio requires you to keep a reference to a Task while it is running and only uses weakrefs to the Tasks internally.

**Reply by SandyWalsh** [reply]
agreed ... this is the answer I'm going with.

**Reply by peterhinch** [reply]
> This is a difference between Micropython and Python - Python Asyncio requires you to keep a reference to a Task while it is running and only uses weakrefs to the Tasks internally.

This is not entirely true. Please see [this ref](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md#note-on-cpython-compatibility).

tl;dr CPython requires application code to maintain an explicit reference to a task for its duration. MP allows this but does not require it. In my view this is a nasty feature* of CPython: if you fail to create a reference, rare RAM issues may occur.

* bug
